### DISCUSSION TITLE
Adding arguments to Pool

### DISCUSSION THREAD

**Discussion by ikelaiah** [mr-highball/ezthreads#21] [General]
Hi @mr-highball ,

Let say, I have 1000 lines of string, and would like to get each thread in the the pool to process 100 each.
For the first 100 lines of string, I guess, I need to do the following.

```.pascal
ThreadPool
      .AddArg('lines', first_100_lines)
      .Queue(ProcessLines, nil, nil);
      .Start;
```

How do I tell the other threads in the pool to process the remainig?
Or perhaps, I am approaching the problem incorrectly? I'd welcome suggestions.

Thank you.

**Comment by mr-highball** [comment]
Hi there @ikelaiah -

Instead of directly adding the string chunks into to the pool, I'd probably setup a collection and a critical section that the worker threads would use to pull from

You could call `.AddArg('lines', Integer(Pointer(TStringList.Create()))`

And then -

`.AddArg('critical', Integer(Pointer(TCritical.Create()))`

From there your process function could first

* enter the critical section
* Check to see if lines has any records
* extract a record (string chunk)
* exit critical
* do it's work

To _feed_ the thread pool, some external function could either pre-populate the string list with chunks or could access the lines variable and feed it as it needs, allowing the workers to pick up work when it arrives 


Also I'm on my phone so sorry if there's typos, but the general concept should work for you


Links:
* https://wiki.lazarus.freepascal.org/Multithreaded_Application_Tutorial#Critical_sections

**Reply by ikelaiah** [reply]
I got the gist, and thanks for pointing out the Critical Section too.
Thank you @mr-highball !
